trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/NwaCegarLoop.java

```diff
@@ -574,4 +576,15 @@ private static SearchStrategy getSearchStrategy(final IPreferenceProvider mPrefs
 			throw new IllegalArgumentException();
 		}
 	}
+
+	private List<IcfgLocation> getIcfgLocationsFromRun(final NestedRun<L, IPredicate> run) {
+		return run.getStateSequence().stream().map(p -> ((ISLPredicate) p).getProgramPoint())
+				.collect(Collectors.toList());
```

Comment:
The point was to not hardcode such logic deep inside a trace check, which has no idea what the control configurations of the original program look like, and thus did the wrong thing for some of our CEGAR loops.

On the other hand, the CEGAR loops know about the control configurations of the program they are verifying, so they can provide sensible, specialized implementations.

Reply:
Ok, that makes perfect sense. Nevertheless, why is it not a `NestedRun<L, ISLPredicate>`?

Reply:
Because the `mCounterexample` run created by `NwaCegarLoop` is of type `NestedRun<L, IPredicate>` (as set by `::isAbstractionEmpty`).

Reply:
Because the type of states is not necessarily `ISLPredicate` -- this CEGAR loop can also be used with concurrent programs, just not the danger invariant support (to which this method belongs).

Reply:
Yes, so you are saying, we would have to propagate the type too far up?

Reply:
Yes. Basically, moving it up would replace the (checked) cast here with an unchecked cast further up, which I want to avoid.